Implementation of the StableDag data structure

davenza opened the pull request:
This is an early implementation of a StableDag.The code is almost copied from the Dag structure, but using StableGraph instead of Graph.  As indicated in the original commit, there are 4 methods that are not implemented because StableGraph does not provide the necessary functions:

    raw_nodes() -> Because StableGraph lacks raw_nodes().
    node_weights_mut() -> Because StableGraph lacks node_weights_mut().
    raw_edges() -> Because StableGraph lacks raw_edges().
    edge_weights_mut() -> Because StableGraph lacks edge_weights_mut().

I included the same set of tests for StableDag that is used in the Dag structure in the files
tests/add_edges_stable.rs
tests/iterators_stable.rs
tests/walkers_stable.rs

I also included the tests in tests/stable_indices to test that node/edge indices do not change when we remove an edge/node.

I am pretty new to Rust, so it may need some better code / project organization.

davenza committed:
Implementation of the StableDag data structure. This data structure is based on the StableGraph data structure in petgraph. StableDag contains almost the same functionality as Dag. However, the following methods are not implemented because StableGraph in petgraph does not provide the necessary methods:

raw_nodes() -> Because StableGraph lacks raw_nodes().
node_weights_mut() -> Because StableGraph lacks node_weights_mut().
raw_edges() -> Because StableGraph lacks raw_edges().
edge_weights_mut() -> Because StableGraph lacks edge_weights_mut().

mitchmindtree commented on src/stabledag.rs:
```diff
@@ -0,0 +1,936 @@
+        self.graph.edge_weight_mut(edge)
+    }
+
+    /// Read from the internal edge array.
```
We probably don't need these commented methods - we likely won't ever implement them as I believe the way edges are stored differs between the `Graph` and `StableGraph`.

mitchmindtree commented on src/stabledag.rs:
```diff
@@ -0,0 +1,936 @@
+/// An error returned by the `StableDag::add_edge` method in the case that adding an edge would have
+/// caused the graph to cycle.
+#[derive(Copy, Clone)]
+pub struct WouldCycle<E>(pub E);
```
I think I'd prefer if we re-used the `WouldCycle` error from the crate root rather than declaring a new one for `StableDag`. It might just make error handling downstream a bit easier in cases where the user is working with both `Dag`s and `StableDag`s.

mitchmindtree commented on src/stabledag.rs:
```diff
@@ -0,0 +1,936 @@
+impl<'a, N, E, Ix> IntoNodeIdentifiers for &'a StableDag<N, E, Ix>
+    where
+        Ix: IndexType
+
```
unnecessary space

mitchmindtree reviewed (changes requested):
Thanks so much @davenza, this is really great stuff! I've been meaning to get around to this for a while, I can definitely see myself using this down the track.

I just have a couple very minor things to clean up - otherwise I'm more than happy to merge!

So sorry for the delay by the way, it's getting a little overwhelming maintaining what must be ~50 crates by now, whilst also working at my own business! Thanks for being patient with me :)

davenza committed:
- The stabledag module uses the root WouldCycle.
- Removed commented methods not implemented by petgraph.
- Re-formated the stabledag.rs file to have the same format as lib.rs:
	- In the where clauses, it uses the same space formatting as in lib.rs.
	- Re-formated some comments to fit in 100 columns.
- Corrected typo in the description of the module.

davenza committed:
Added a "stable_dag" feature for conditional compilation in a similar way to the "stable_graph" feature in petgraph.

davenza commented:
Hi, thanks for you review. I made the changes you asked for. They are in 5482078. 

Then, I added a feature called `stable_dag` for conditional compilation in cf752f4. What do you think about this?

davenza committed:
Fix: Include conditional compilation for the tests.

davenza commented:
I forgot to include conditional compilation in the tests, but it should be solved in cce14bc.

davenza reviewed (commented):
Reviewed changes.

davenza committed:
Implement the From trait from Dag to StableDag.

davenza committed:
Added contains_node() as in StableGraph.

fwcd commented:
Any updates on this PR? I would love to help if there are any specific issues, especially in regard to https://github.com/RustAudio/dsp-chain/issues/103.

mitchmindtree commented:
Thanks for the reminder @fwcd!
